azkaban-exec-server/src/main/java/azkaban/dag/Node.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.dag;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Node in a DAG: Directed acyclic graph.
+ */
+class Node {
+
+  private final String name;
+
+  private final NodeProcessor nodeProcessor;
+
+  // The nodes that this node depends on.
+  private final List<Node> parents = new ArrayList<>();
+
+  // The nodes that depend on this node.
+  private final List<Node> children = new ArrayList<>();
+
+  private Status status = Status.READY;
+
+  private Flow flow;
+
+  Node(final String name, final NodeProcessor nodeProcessor) {
+    this.name = name;
+    requireNonNull(nodeProcessor);
+    this.nodeProcessor = nodeProcessor;
+  }
+
+  public Flow getFlow() {
+    return this.flow;
+  }
+
+  public void setFlow(final Flow flow) {
+    this.flow = flow;
+  }
+
+  private Node addParent(final Node node) {
+    this.parents.add(node);
+    return this;
+  }
+
+  Node addChild(final Node node) {
+    this.children.add(node);
+    node.addParent(this);
+    return this;
+  }
+
+  Node addChildren(final Node... nodes) {
+    for (final Node node : nodes) {
+      addChild(node);
+    }
+    return this;
+  }
+
+  boolean hasParent() {
+    return !this.parents.isEmpty();
+  }
+
+  boolean isReady() {
+    if (this.status != Status.READY) {
+      return false;
+    }
+    for (final Node parent : this.parents) {
+      if (!parent.isSuccessEffectively()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public void run() {
+    changeStatus(Status.RUNNING);
+  }
+
+  void markSuccess() {
+    changeStatus(Status.SUCCESS);
+    for (final Node child : this.children) {
+      child.check();
+    }
+    this.flow.checkFinished();
+  }
+
+  /**
+   * Checks if all the dependencies are met and run if they are.
+   */
+  private void check() {
+    if (this.status == Status.DISABLED) {
+      return;
+    }
+
+    for (final Node node : this.parents) {
+      if (!node.isSuccessEffectively()) {
+        return;
+      }
+    }
+
+    run();
+  }
+
+  /**
+   * @return true if the dependency requirement is considered satisfied
```

Comment:
Maybe add to the end of that sentence: "..for the nodes that depend on this node". Without that it seems a bit confusing, I guess because of the word "dependency".

Reply:
Removed this method.